File: programmers_problems/92334.py
# 신고 결과 받기

import collections

"""
해시 테이블을 사용하여 접근한 이유

우선 사람에 따른 신고자들을 저장하기 위해서 key, value 형식으로 저장할 필요가 있어 해시 테이블을 사용했다
또한, 신고에 성공한 사람들도 기록하기 위해 이 또한 해시 테이블을 사용했다
"""


def solution(id_list, report, k):
    report_dic = collections.defaultdict(list)
    report_num = collections.defaultdict(int)
    
    # id 리스트 기록
    for id in id_list:
        report_dic[id]
        report_num[id]
        
    # 신고 기록 
    for value in report:
        # 신고한 사람, 신고받은 사람
        from_person, to_person = value.split(" ")
        # 신고한 사람이 같은 사람을 두 번 이상 중복 신고하지 않았다면
        if from_person not in report_dic[to_person]:
            report_dic[to_person].append(from_person)
            
    for _, value in report_dic.items():
        # k번 이상 신고 당함
        if len(value) >= k:
            for from_person in value:
                report_num[from_person] += 1
                
    return list(report_num.values())

File: programmers_problems/92334.kt
package programmers

/**
 * 이용자의 ID와 신고한 이용자의 ID 정보가 담긴 문자열 배열이 주어지므로
 * 이를 map을 통해 저장하는데, 하나는 이용자의 ID를 key로하고 신고 성공 횟수를 value로 하여 선언하고,
 * 하나는 이용자의 ID를 key로하고 해당 ID를 신고한 사람들을 value로 설정한다.
 * 이제 이용자의 ID에 그 ID를 신고한 사람들이 기록되어 있으므로 신고횟수가 K개 이상이라면
 * 신고에 성공한 사람들의 Count를 증가시키는 방식으로 풀이했다.
 */

class Solution {
    fun solution(id_list: Array<String>, report: Array<String>, k: Int): IntArray {
        val reportDic = mutableMapOf<String, MutableSet<String>>()
        val reportNum = mutableMapOf<String, Int>()

        for(id in id_list) {
            reportDic[id] = mutableSetOf()
            reportNum[id] = 0
        }

        for(value in report) {
            val (fromPerson, toPerson) = value.split(" ")
            reportDic[toPerson]?.add(fromPerson)
        }

        for(value in reportDic.values) {
            if(value.size >= k) {
                for(fromPerson in value) {
                    reportNum[fromPerson] = reportNum[fromPerson]!!.plus(1)
                }
            }
        }

        return reportNum.values.toIntArray()
    }
}

fun main() {
    val sol = Solution()
    val idList = arrayOf("muzi", "frodo", "apeach", "neo")
    val reportList = arrayOf("muzi frodo","apeach frodo","frodo neo","muzi neo","apeach muzi")
    val k = 2

    val result = sol.solution(idList, reportList, k)
    for(x in result) print("$x ")
}